src/leap/bitmask/core/dispatcher.py

```diff
@@ -0,0 +1,196 @@
+# -*- coding: utf-8 -*-
+# dispatcher.py
+# Copyright (C) 2016 LEAP
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Command dispatcher.
+"""
+import json
+
+from twisted.internet import defer
+from twisted.python import failure, log
+
+
+# TODO implement sub-classes to dispatch subcommands (user, mail).
+
+
+class CommandDispatcher(object):
+
+    def __init__(self, core):
+
+        self.core = core
+
+    def _get_service(self, name):
+
+        try:
+            return self.core.getServiceNamed(name)
+        except KeyError:
+            return None
+
+    def dispatch(self, msg):
+        cmd = msg[0]
+
+        _method = getattr(self, 'do_' + cmd.upper(), None)
+
+        if not _method:
+            return defer.fail(failure.Failure(RuntimeError('No such command')))
+
+        return defer.maybeDeferred(_method, *msg)
+
+    def do_STATS(self, *parts):
+        return _format_result(self.core.do_stats())
+
+    def do_VERSION(self, *parts):
+        return _format_result(self.core.do_version())
+
+    def do_STATUS(self, *parts):
+        return _format_result(self.core.do_status())
+
+    def do_SHUTDOWN(self, *parts):
+        return _format_result(self.core.do_shutdown())
+
+    def do_USER(self, *parts):
+
+        subcmd = parts[1]
+        user, password = parts[2], parts[3]
+
+        bf = self._get_service('bonafide')
+
+        if subcmd == 'authenticate':
+            d = bf.do_authenticate(user, password)
+
+        elif subcmd == 'signup':
+            d = bf.do_signup(user, password)
+
+        elif subcmd == 'logout':
+            d = bf.do_logout(user, password)
+
+        elif subcmd == 'active':
+            d = bf.do_get_active_user()
+
+        d.addCallbacks(_format_result, _format_error)
+        return d
+
+    def do_EIP(self, *parts):
```

Comment:
actually, I think it might not be needed. when we get the service, it will return None, ie, disabled.